### ISSUE TITLE
Homebridge integration

### ISSUE THREAD

**Issue by xuncia** [alanreid/bravia#4]
Hi,
I was wondering if it will be possibile to integrate your code with homebridge(https://github.com/nfarina/homebridge) as an accessory or plugin.
Do you have any hint in this case?
Anycase tonight I'll try to take a look to get it work, so I can give command to my bravia through Siri.

**Comment by alanreid** [comment]
Hi,

I have no idea, but that would be awesome! 
Let me know if you can make it work, I'd gladly help.

Best,
Alan

**Comment by druchtie** [comment]
@Xuncia that would be great! I also gladly would 
like to help!

**Comment by xuncia** [comment]
I spent 30 minutes last night but I had some problems with the code from my bravia (KDL-50W829B), it was continuing asking me the code for each command I sent....let's see if in these christmas holidays my children will give me more time :) I let you know.

**Comment by alanreid** [comment]
Hm, that's odd. Check if the cookiejar file is being created and has write permissions. Can't think of anything else :-/

**Comment by druchtie** [comment]
I'm having the same problem.  I can authorise my tv (4 digit's) and the cookie (content) is generated. But he will not execute the command. Instead he will ask me for another 4 digit code. I think he is not reading the cookie. I have changed the permission to 777 but it doesn't seem to work.  Any thoughts?

if(!this.hasCookie()) {
    console.log('no cookies');

returns no cookies :(

**Comment by alanreid** [comment]
Could you send me the contents of your `cookies.json` file?

Maybe the domain of the cookie is different, this is the way it's checked:

```
Bravia.prototype.hasCookie = function() {
  return this.cookieJar.getCookies('http://' + this.ip + '/sony').length > 0;
};
```

**Comment by druchtie** [comment]
{"192.168.178.15":{"/sony/":{"auth":{"key":"auth","value":"7f1cc53a5d58e6e9660b841f5a4a24eab921e51778df6efb26c895e3beed2346","expires":"2016-01-11T11:42:45.000Z","maxAge":1209600,"domain":"192.168.178.15","path":"/sony/","hostOnly":true,"creation":"2015-12-28T10:53:07.643Z","lastAccessed":"2015-12-28T11:42:47.052Z"}}}}

**Comment by alanreid** [comment]
Thanks. Could you try adding a slash after "/sony" [here](https://github.com/alanreid/bravia/blob/master/lib/index.js#L167)?  
Like so: `return this.cookieJar.getCookies('http://' + this.ip + '/sony/').length > 0;`

If that works, I'll patch this to check for both cases (with and without the slash)

**Comment by xuncia** [comment]
No alanreid it doesn't work, it doesn't exec the command and at the next call ask for a new code...

**Comment by alanreid** [comment]
Does the IP in `cookies.json` match the one you set for the TV?

**Comment by xuncia** [comment]
yes, I was thinking the same, but it's the same.

**Comment by alanreid** [comment]
I'm now seeing that [tough-cookie](https://github.com/SalesforceEng/tough-cookie/blob/master/lib/cookie.js#L1022)'s `getCookies()` method is async now. 

Could you try replacing `getCookies` with `getCookiesSync` please? Sorry I can't test it myself, I moved and I don't have a Bravia TV anymore :(

**Comment by xuncia** [comment]
I'll try in a few minutes and I let you know

**Comment by xuncia** [comment]
return this.cookieJar.getCookiesSync('http://' + this.ip + '/sony/').length 
                        ^
TypeError: undefined is not a function
    at Bravia.hasCookie (/Users/luciano/Downloads/bravia-master/lib/index.js:167:25)
    at Bravia.auth (/Users/luciano/Downloads/bravia-master/lib/index.js:174:12)
    at new Bravia (/Users/luciano/Downloads/bravia-master/lib/index.js:21:10)
    at module.exports (/Users/luciano/Downloads/bravia-master/lib/index.js:227:10)
    at Object.<anonymous> (/Users/luciano/Downloads/bravia-master/demo.js:4:1)
    at Module._compile (module.js:460:26)
    at Object.Module._extensions..js (module.js:478:10)
    at Module.load (module.js:355:32)
    at Function.Module._load (module.js:310:12)
    at Function.Module.runMain (module.js:501:10)

**Comment by alanreid** [comment]
Ug... no then... 

does this return something? `this.cookieJar.getCookies('http://' + this.ip + '/sony/')`

**Comment by xuncia** [comment]
yep...I'll post what is my output from the shell

node demo.js
Please enter the 4-digit code shown on your TV: 0049
Please enter the 4-digit code shown on your TV: PowerOn, PowerOff, Input, GGuide, EPG, Favorites, Display, Home, Options, Return, Up, Down, Right, Left, Confirm, Red, Green, Yellow, Blue, Num1, Num2, Num3, Num4, Num5, Num6, Num7, Num8, Num9, Num0, Num11, Num12, VolumeUp, VolumeDown, Mute, ChannelUp, ChannelDown, SubTitle, ClosedCaption, Enter, DOT, Analog, Teletext, Exit, Analog2, *AD, Digital, Analog?, BS, CS, BSCS, Ddata, PicOff, Tv_Radio, Theater, SEN, InternetWidgets, InternetVideo, Netflix, SceneSelect, Mode3D, iManual, Audio, Wide, Jump, PAP, MyEPG, ProgramDescription, WriteChapter, TrackID, TenKey, AppliCast, acTVila, DeleteVideo, PhotoFrame, TvPause, KeyPad, Media, SyncMenu, Forward, Play, Rewind, Prev, Stop, Next, Rec, Pause, Eject, FlashPlus, FlashMinus, TopMenu, PopUpMenu, RakurakuStart, OneTouchTimeRec, OneTouchView, OneTouchRec, OneTouchStop, DUX, FootballMode, Social

after this nothing, after a press something + return I'm going back to the shell, if I try again I'll be asked for a new code

**Comment by alanreid** [comment]
Could you post the output of

```
Bravia.prototype.hasCookie = function() {
  console.log(this.cookieJar.getCookies('http://' + this.ip + '/sony'));
  return this.cookieJar.getCookies('http://' + this.ip + '/sony').length > 0;
};
```

? 
(note the console.log)

I'm running out of ideas :S

**Comment by xuncia** [comment]
it seems it doesn't get the cookies....

VirgolaServer2:bravia-master luciano$ node demo.js
[]
Please enter the 4-digit code shown on your TV: 1992
[]
Please enter the 4-digit code shown on your TV: PowerOn, PowerOff, Input, GGuide, EPG, Favorites, Display, Home, Options, Return, Up, Down, Right, Left, Confirm, Red, Green, Yellow, Blue, Num1, Num2, Num3, Num4, Num5, Num6, Num7, Num8, Num9, Num0, Num11, Num12, VolumeUp, VolumeDown, Mute, ChannelUp, ChannelDown, SubTitle, ClosedCaption, Enter, DOT, Analog, Teletext, Exit, Analog2, *AD, Digital, Analog?, BS, CS, BSCS, Ddata, PicOff, Tv_Radio, Theater, SEN, InternetWidgets, InternetVideo, Netflix, SceneSelect, Mode3D, iManual, Audio, Wide, Jump, PAP, MyEPG, ProgramDescription, WriteChapter, TrackID, TenKey, AppliCast, acTVila, DeleteVideo, PhotoFrame, TvPause, KeyPad, Media, SyncMenu, Forward, Play, Rewind, Prev, Stop, Next, Rec, Pause, Eject, FlashPlus, FlashMinus, TopMenu, PopUpMenu, RakurakuStart, OneTouchTimeRec, OneTouchView, OneTouchRec, OneTouchStop, DUX, FootballMode, Social

**Comment by alanreid** [comment]
The issue might be with the [cookie store](https://github.com/alanreid/bravia/blob/master/lib/index.js#L17) then. Could you try updating it to 0.0.2? Otherwise I'll have to take a deeper look at that tonight.

**Comment by xuncia** [comment]
I just updated with the last version of file-cookie-store with npm, but nothing changed... :(

**Comment by alanreid** [comment]
I'll take a look at this tonight then

**Comment by xuncia** [comment]
I think that I figure out....it was a problem about the path the cookie is in system..but still the command doesn't work...I'l try tonight...

**Comment by corpr8** [comment]
Just a suggestion - try my open PR and see if that fixes your issue...

**Comment by druchtie** [comment]
Awesome! The suggestion of @corpr8 did the trick. 
changed: options.jar = this.cookieJar; 
to: options.jar = true; //has cookie 

Now I need to figure out why I need to authenticate every time. 

@corpr8 could share your index.js file? I din't get the options.auth part

**Comment by corpr8** [comment]
I've been thinking about that. This.cookieJar should be persisting these cookies out to disk (which it is, you can open the file and see them if you set the permissions correctly).

So I am thinking that the above is just a hack and the real problem is getting the requests to use the saved cookie. Will keep playing.

**Comment by alanreid** [comment]
Yeah, the cookiejar store is probably the issue here. We could try with another implementation. Unfortunately I can't test this live, so I'll try to reproduce this by extracting that part of the code and see what's going on.

**Comment by corpr8** [comment]
Reading through tough-cookie - doesn't .getCookies need a callback rather than being async?

**Comment by lombardi-zz** [comment]
This is my code. Is not clean because I have no time. It based on "home bridge-sonytvpower".
Replace index.js with attached code.

Install also with npm "base-64","ping" and"prompt".
1. Start home bridge manually on the Console
2. Get the pin from TV and enter it on Console 

Channel 1001 Starts "Fireplace Screensaver" 

I hope thats help.

[Archiv.zip](https://github.com/alanreid/bravia/files/184993/Archiv.zip)

**Comment by alanreid** [comment]
Hi @lombardi-zz 

I'll take a look and see how this one's implemented. If I find something promising, I'll add it to the lib and have you guys test it.

Thanks!

**Comment by alisdairjsmyth** [comment]
@lombardi-zz - are you planning to publish a homebridge plugin based on the code you attached in the comment above?  homebridge-sonytvpower has too limited a scope.

(I have been looking for such a plugin and have been considering writing one myself.)

**Comment by lombardi-zz** [comment]
i cleaning my code.

you can install it with `npm install -g homebridge-sonytvremote`

[Ver02.zip](https://github.com/alanreid/bravia/files/190976/Ver02.zip)

**Comment by alanreid** [comment]
Hi there,

I got a Bravia TV for a week again and I was able to reproduce and fix the issue. 
The problem was with the tough-cookie package (`tough-cookie/lib/cookie.js:576`), which replaces the tailing slash only on lookup and not on save. This means it's not able to find a cookie that's saved as `/sony/`, because it will always look it up as `/sony`. 

So pull the latest changes and it should work again.

Hope it helps!
Alan
